lsif-semanticdb/src/main/java/com/sourcegraph/lsif_semanticdb/LsifSemanticdb.java

```
@@ -0,0 +1,132 @@
+    return options.parallel ? files.parallelStream() : files.stream();
+  }
+
+  private Set<String> exportSymbols(List<Path> files) {
+    return filesStream(files)
+        .flatMap(this::parseTextDocument)
+        .flatMap(
+            d ->
+                d.semanticdb.getOccurrencesList().stream()
+                    .filter(occ -> occ.getRole() == Role.DEFINITION)
+                    .map(SymbolOccurrence::getSymbol)
+                    .filter(SemanticdbSymbols::isGlobal))
+        .collect(Collectors.toSet());
+  }
+
+  private Stream<Integer> processPath(Path path, Set<String> isExportedSymbol) {
+    return parseTextDocument(path).map(d -> processDocument(d, isExportedSymbol));
+  }
+
+  private Integer processDocument(LsifTextDocument doc, Set<String> isExportedSymbol) {
+    int documentId = writer.emitDocument(doc);
+    Set<String> localDefinitions =
+        doc.semanticdb.getOccurrencesList().stream()
+            .filter(
+                occ ->
+                    occ.getRole() == Role.DEFINITION && SemanticdbSymbols.isLocal(occ.getSymbol()))
+            .map(SymbolOccurrence::getSymbol)
+            .collect(Collectors.toSet());
+    doc.id = documentId;
+    ResultSets results = new ResultSets(writer, globals, isExportedSymbol, localDefinitions);
+    Set<Integer> rangeIds = new LinkedHashSet<>();
+
+    for (SymbolOccurrence occ : doc.sortedSymbolOccurrences()) {
+      SymbolInformation symbolInformation =
+          doc.symbols.getOrDefault(occ.getSymbol(), SymbolInformation.getDefaultInstance());
+      ResultIds ids = results.getOrInsertResultSet(occ.getSymbol());
+      int rangeId = writer.emitRange(occ.getRange());
+      rangeIds.add(rangeId);
+
+      // Range
+      writer.emitNext(rangeId, ids.resultSet);
+
+      // Reference
+      writer.emitItem(ids.referenceResult, rangeId, doc.id);
+
+      // Definition
+      if (occ.getRole() == SymbolOccurrence.Role.DEFINITION && ids.isDefinitionDefined()) {
```

in what case is `ids.isDefinitionDefined() == true` aka when would we have two identical definitions?

---

The definition ID should always be available in this case, it's a bug if it's not. I updated the code to report an error message in this case instead of silently ignoring it.